[CDU] F-PLN page visual rework

Lollo999 opened the pull request:
## Summary of Changes
This PR is a follow up to the FPLN scroll fix #2069 
Made some changes to the F-PLN page to be more accurate to the real one:
1> **FIXED** : Scroll skipping the "END OF F-PLN" line when scrolling
2> **FIXED** : "**SPD/ALT**" indication is now fixed on top, making it easier to read the data
3> **FIXED** : Destination airport line is now fixed on the bottom of the page (if ERASE/INSRT line is not present)
4> **ADDED** : Unit of measure **[NM]** on the last row waypoint distance from previous point.
5> **FIXED** : Text size and color in the first and last point **(DEP/DEST)**
6> **FIXED:** "From" text should only be present if the wp the aircraft is flying from is in the first row of the page
7> **FIXED**: Now also the departure and arrival wp in the flight plan change color when the plan is being edited
8> **FIXED**: Repetition of waypoints when less than 5 wp are inserted
9> **ADDED**: PPOS is now in green
10> **EDIT** : changed page header from "[ICAO][RWY]" to F-PLN, as the first notation may be confusing during flight and i found no evidence of it in the honeywell docs
11> Minor edits and code clean up

p.s. had to refork the repo and make a new PR, and also i'll provide better screens and gifs

EDIT----------------------------------------------------
New features added, screens in the comments:
12>**FIXED** : Text alignment for SPD/ALT data
13>**FIXED** : Removed TMPY from ERASE/INSERT*
14>**FIXED** : SPD/ALT colors
15>**FIXED** : "NM" measure unit moved to the second row instead of last one
16>**ADDED** : Runway altitude for DEP and ARR runways


## Screenshots (if necessary)
**BEFORE**:

On flight plan **insert**:
![Microsoft Flight Simulator Screenshot 2020 12 02 - 07 58 21 52 (2)](https://user-images.githubusercontent.com/36545051/101472214-94189680-3948-11eb-8bbd-36eb78d1c9d8.png)

**Editing** flight plan:
![Microsoft Flight Simulator Screenshot 2020 12 08 - 10 55 35 72 (2)](https://user-images.githubusercontent.com/36545051/101472352-ccb87000-3948-11eb-93d1-9374cbe4ab1a.png)

Scrolling:
![Microsoft Flight Simulator 2020 12 08 - 10 58 29 04_Trim](https://user-images.githubusercontent.com/36545051/101473039-a21ae700-3949-11eb-9d09-a1eaa005e2cc.gif)


**AFTER:**

on flight plan **insert**:
![Microsoft Flight Simulator Screenshot 2020 12 08 - 11 06 27 37 (2)](https://user-images.githubusercontent.com/36545051/101472444-e78ae480-3948-11eb-8ae0-0d3dc1cc666d.png)

**Editing** flight plan:
![Microsoft Flight Simulator Screenshot 2020 12 08 - 11 07 32 44 (2)](https://user-images.githubusercontent.com/36545051/101472498-f8d3f100-3948-11eb-8117-313a78163c14.png)

**Scrolling**:
![Microsoft Flight Simulator 2020 12 08 - 11 08 57 05_Trim](https://user-images.githubusercontent.com/36545051/101473008-99c2ac00-3949-11eb-8e7e-50b4606aab46.gif)

## References
Honeywell manual, online videos and compared by visual clues to the FSL 320 for some behaviour details

## Additional context
The update tested great so far, further feedback and future improvement suggestions are very appreciated


Discord username (if different from GitHub): Lorenzo Pinna#9154

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

beheh commented:
Great job! I have a few more small things for this page. Generally note that FSLabs implements Thales, not Honeywell, so there is some small discrepancies:

1. `TMPY ERASE` / `TMPY INSERT` are Thales button labels - could you maybe change that to `{ERASE` and `INSERT*`?

![image](https://user-images.githubusercontent.com/929769/101477172-27ed6100-394f-11eb-9820-34770e324f57.png)

2. The NM append should be on the second line, not on the second-to-last one. Compare this screenshot of Honeywell on the left, and FSLabs with Thales on the right:

![image](https://user-images.githubusercontent.com/929769/101477296-553a0f00-394f-11eb-9796-4f42fe831844.png)

3. Also note that in the screenshot above "NM" is the same color as the number - so green normally, and yellow in TMPY.

And one more note:

* Technically there's no header at all, not even F-PLN - the real aircraft has a split header, but we don't support that yet:
![image](https://user-images.githubusercontent.com/929769/101477036-ef4d8780-394e-11eb-80f2-8d7da2c2766e.png)
As shown above also, that header is replaced with `TMPY` on the secondary flight plan.

--

Here's the references I used:

Honeywell left, FSLabs with Thales on the right: https://www.youtube.com/watch?v=twwUjHXdNVU
Full Honeywell flight prep: https://www.youtube.com/watch?v=D7HjVlvKd8o

Lollo999 commented:
> Great job! I have a few more small things for this page. Generally note that FSLabs implements Thales, not Honeywell, so there is some small discrepancies:
> 
> 1. `TMPY ERASE` / `TMPY INSERT` are Thales button labels - could you maybe change that to `{ERASE` and `INSERT*`?
> 
> ![image](https://user-images.githubusercontent.com/929769/101477172-27ed6100-394f-11eb-9820-34770e324f57.png)
> 
> 1. The NM append should be on the second line, not on the second-to-last one. Compare this screenshot of Honeywell on the left, and FSLabs with Thales on the right:
> 
> ![image](https://user-images.githubusercontent.com/929769/101477296-553a0f00-394f-11eb-9796-4f42fe831844.png)
> 
> 1. Also note that in the screenshot above "NM" is the same color as the number - so green normally, and yellow in TMPY.
> 
> And one more note:
> 
> * Technically there's no header at all, not even F-PLN - the real aircraft has a split header, but we don't support that yet:
>   ![image](https://user-images.githubusercontent.com/929769/101477036-ef4d8780-394e-11eb-80f2-8d7da2c2766e.png)
>   As shown above also, that header is replaced with `TMPY` on the secondary flight plan.
> 
> --
> 
> Here's the references I used:
> 
> Honeywell left, FSLabs with Thales on the right: https://www.youtube.com/watch?v=twwUjHXdNVU
> Full Honeywell flight prep: https://www.youtube.com/watch?v=D7HjVlvKd8o

Thank you for the feedback, 
i double checked in the manuals, i'll try to implement these changes asap, 
Yep noticed the F-PLN shouldn't be there at all, but when i tried to remove it i broke everything so i guess it works as an index value for the page of some sort.

Lollo999 commented:
Added more feature to this PR
>**FIXED** : Text alignment for SPD/ALT data
>**FIXED** : removed TMPY from ERASE/INSERT*
>**FIXED** : SPD/ALT colors
>**FIXED**  : "NM" measure unit moved to the second row instead of last one
>**ADDED** : Runway altitude for DEP and ARR runways
-----------------------------------------------------------------------------------
**NEW SCREENS**

![Microsoft Flight Simulator Screenshot 2020 12 09 - 15 56 41 36 (2)](https://user-images.githubusercontent.com/36545051/101647574-2d26da80-3a39-11eb-94b2-e355051c2652.png)
![Microsoft Flight Simulator Screenshot 2020 12 09 - 15 55 33 29 (2)](https://user-images.githubusercontent.com/36545051/101647591-32842500-3a39-11eb-9fc1-f2824749f755.png)


**SCROLLING** 

![Microsoft Flight Simulator 2020 12 09 - 15 54 26 01_Trim (1)](https://user-images.githubusercontent.com/36545051/101647968-a3c3d800-3a39-11eb-88a8-6c0aea45571a.gif)

Benjozork commented on .github/CHANGELOG.md:
```diff
@@ -141,6 +141,8 @@
 1. [PFD] Changed S and F speeds data source - @MisterChocker (Leon)
 1. [ANIMATIONS] Desynced the wiper animations - @pepperoni505 (pepperoni505), @lukecologne (luke)
+1. [CDU] F-PLN page visual rework - @Lollo999 (Lorenzo Pinna)
+
```
Remove extra line

```suggestion
```

Benjozork replied:
Remove extra line
```suggestion
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -13,9 +13,10 @@ class CDUFlightPlanPage {
         const isFlying = Simplane.getAltitudeAboveGround() > 10 ||
             Simplane.getEngineThrottleMode(0) >= ThrottleMode.FLEX_MCT && Simplane.getEngineThrottleMode(1) >= ThrottleMode.FLEX_MCT;
         let originIdentCell = "----";
+        var runway = null;
```
Don't use `var`. Use `let`.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -31,19 +32,20 @@ class CDUFlightPlanPage {
             }
         }
         let destCell = "----";
+        var approachRunway = null;
```
Same here.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -124,17 +151,25 @@ class CDUFlightPlanPage {
             }
             const index = iWaypoint;
             iWaypoint++;
+            var depAltColor = "white";
+            var depAlt = "-----";
```
Same here

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -143,9 +178,16 @@ class CDUFlightPlanPage {
+            } else if (index === waypointsWithDiscontinuities.length - 1) {
                 let destTimeCell = "----";
                 let destDistCell = "---";
+                var apprElev = "-----";
+                var apprColor = "white";
```
And here

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -143,9 +178,16 @@ class CDUFlightPlanPage {
+                    apprElev = (approachRunway.elevation * 3.280).toFixed(0).toString();
+                    apprColor = color;
+                }
+                apprElev = apprElev.padStart(6,"\xa0");
```
```suggestion
                apprElev = apprElev.padStart(6, "\xa0");
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -224,53 +269,59 @@ class CDUFlightPlanPage {
                         }
                         const distance = (waypoint === mcdu.flightPlanManager.getActiveWaypoint() ? waypoint.liveDistanceTo : waypoint.distanceInFP);
-                        rows[2 * i] = [airwayName, (index >= activeIndex || waypoint.ident === "(DECEL)" ? distance.toFixed(0) : "")];
+                        var dstnc;
```
Same

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -224,53 +269,59 @@ class CDUFlightPlanPage {
+                        } else {
+                            dstnc = distance.toFixed(0).toString() + "\xa0\xa0";
+                        }
+                        for (var z = 0; z < 9 - dstnc.length; z++) {
```
Same

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -324,16 +375,26 @@ class CDUFlightPlanPage {
                             }
                         } else {
-                            rows[2 * i + 1] = [waypoint.ident + "[color]" + color, speedConstraint + "/" + altPrefix + altitudeConstraint + "[s-text][color]" + color, timeCell + "[color]" + color];
+                            var spdColor = color;
+                            var altColor = color;
```
Same

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -369,24 +428,28 @@ class CDUFlightPlanPage {
-            } else {//else go to the bottom
-                offset = waypointsWithDiscontinuities.length - 1;
+            if (waypointsWithDiscontinuities.length > 4) {//scroll only if there are more than 5 points
+                if (offset > 0) {//if page not on top
+                    offset--;
+                } else {//else go to the bottom
```
```suggestion
                if (offset > 0) { // if page not on top
                    offset--;
                } else { // else go to the bottom
```

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_FlightPlanPage.js:
```diff
@@ -369,24 +428,28 @@ class CDUFlightPlanPage {
-                offset++;
-            } else {//else go on top
-                offset = 0;
+            if (waypointsWithDiscontinuities.length > 4) {
+                if (offset < waypointsWithDiscontinuities.length) {//if page not on bottom
+                    offset++;
+                } else {//else go on top
+                    offset = 0;
+                }
+                CDUFlightPlanPage.ShowPage(mcdu, offset);
```
```suggestion
            if (waypointsWithDiscontinuities.length > 4) {
                if (offset < waypointsWithDiscontinuities.length) { // if page not on bottom
                    offset++;
                } else { // else go on top
                    offset = 0;
                }
                CDUFlightPlanPage.ShowPage(mcdu, offset);
```

Lollo999 reviewed (commented):
Thanks!

Benjozork reviewed (approved):
LGTM

BritishAvgeek commented:
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: Current date 10/12/2020
Version of the sim: 1.11.7.0
PR Tested: #2287 
Object of test: F-PLAN visual rework

Testing technique: Short EGCC - EGPH
Steps you've made to test the Pull Request:
1. Generated route on Simbrief and validated it checking SkyVector
2. Input into MCDU 
3. Departed and flew the route
4. Monitored usual changes in nm etc during flight and reviewed new rwy alt for departure and arrival runways
5. Landed

Flight notes (If required): N/A
Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: All worked well for me, nice changes thanks

beheh reviewed (approved):
Great job on this!

tjburni commented:
Category: QA Tester
Name: [Z+1]tjburni #0409
Date of testing: Current date 10.12.2020
Version of the sim:1.11.7
PR Tested: #2287
Object of test: F-PLAN visual rework

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. started at gate at eddh
2. input of fpln in mcdu	
3. obspected visuals of cdu while flying
4. flew ifr to eddv
5.landed

Flight notes (If required):

Issues (If Any): none
Overall Rating: Good
Conclusions of the testing:I had no issues. Works ans looks good

armankir commented:
Category: QA Tester
Name: Arman#8145
Date of testing: Current date 12/10/2020
Version of the sim: 1.11.7.0
PR Tested: #2287
Object of test: F-PLAN Page Visual Rework

Testing technique: C&D route entry.

Steps you've made to test the Pull Request:

Spawned into KDTW C&D
Entered flight plan to Chicago O-Hare (KORD)
Looked for any issues in the page.

Flight notes (If required): N/A

Issues (If Any): None
Overall Rating: Good
Conclusions of the testing: Great PR, works like a charm, good job!
